src/kqueue.rs

```
@@ -0,0 +1,380 @@
+            .filter_map(filter_dir)
+        {
+            self.add_single_watch(entry.path().to_path_buf(), is_recursive)?;
+        }
+        self.kqueue.watch()?;
+
+        Ok(())
+    }
+
+    fn add_single_watch(&mut self, path: PathBuf, is_recursive: bool) -> Result<()> {
+        let event_filter = EventFilter::EVFILT_VNODE;
+        let filter_flags = FilterFlag::NOTE_DELETE
+            | FilterFlag::NOTE_WRITE
+            | FilterFlag::NOTE_EXTEND
+            | FilterFlag::NOTE_ATTRIB
+            | FilterFlag::NOTE_LINK
+            | FilterFlag::NOTE_RENAME
+            | FilterFlag::NOTE_REVOKE;
+
+        self.kqueue
+            .add_filename(&path, event_filter, filter_flags)?;
+        self.watches.insert(path, is_recursive);
+        self.kqueue.watch()?;
+        Ok(())
+    }
+
+    fn remove_watch(&mut self, path: PathBuf, remove_recursive: bool) -> Result<()> {
+        match self.watches.remove(&path) {
+            None => return Err(Error::watch_not_found()),
+            Some(is_recursive) => {
+                self.kqueue
+                    .remove_filename(&path, EventFilter::EVFILT_VNODE)
+                    .map_err(|e| Error::io(e))?;
+
+                if is_recursive || remove_recursive {
+                    for entry in WalkDir::new(path)
+                        .follow_links(true)
+                        .into_iter()
+                        .filter_map(filter_dir)
+                    {
+                        self.kqueue.remove_filename(
+                            entry.path().to_path_buf(),
+                            EventFilter::EVFILT_VNODE,
+                        )?;
+                    }
+                }
+                self.kqueue.watch()?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// return `DirEntry` when it is a directory
+fn filter_dir(e: walkdir::Result<walkdir::DirEntry>) -> Option<walkdir::DirEntry> {
+    if let Ok(e) = e {
+        if let Ok(metadata) = e.metadata() {
```

The caller functions returns an error, so I'd like to report permission/access errors here too and not silently drop them

---

This code is copied directly from the INotifyWatcher implementation. So we should resolve this also in there